src/main/java/rx/internal/util/ScalarSynchronousObservable.java

```diff
@@ -41,10 +35,28 @@ public void call(Subscriber<? super T> s) {
                  *  filter it out (such as take(0)). This prevents us from paying the price on every subscription. 
                  */
                 s.onNext(t);
-                s.onCompleted();
+                // checking unsubscribe here because if they really wanted performance then they should have called get().
+                if (!s.isUnsubscribed())
+                    s.onCompleted();
             }
+        };
+
+        ScalarSynchronousObservable<T> o = new ScalarSynchronousObservable<T>(t, onSubscribe);
+
+        // check to see if the execution hooks have wrapped the onSubscribe func
+        if (o.onSubscribe == onSubscribe) {
```

Comment:
As per discussion between @abersnaze and I, we want to explore whether it's possible to have the type define whether the debug hook is synchronous or asynchronous so that we can retain the SynchronousObservable type here. This is so that when tracing is being done, we retain the same code paths through `merge` and others than use the SynchronousObservable optimizations, otherwise observing with the debug hooks will change the execution flow.